src/wp-includes/l10n.php

```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
 		$locale = determine_locale();
 	}
 
-	$mo = new MO();
-	if ( ! $mo->import_from_file( $mofile ) ) {
-		$wp_textdomain_registry->set( $domain, $locale, false );
+	// Ensures the correct locale is set as the current one, in case it was filtered.
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
 
-		return false;
+	/**
+	 * Filters the preferred file format for translation files.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $preferred_format Preferred file format. Possible values: 'php', 'mo'. Default: 'php'.
+	 */
+	$preferred_format = apply_filters( 'translation_file_format', 'php' );
+	if ( ! in_array( $preferred_format, array( 'php', 'mo' ), true ) ) {
+		$preferred_format = 'php';
 	}
 
-	if ( isset( $l10n[ $domain ] ) ) {
-		$mo->merge_with( $l10n[ $domain ] );
+	$mofile_preferred = str_replace( '.mo', ".mo.$preferred_format", $mofile );
+
+	if ( 'mo' !== $preferred_format || str_ends_with( $mofile, $preferred_format ) ) {
+
+		/** This filter is documented in wp-includes/l10n.php */
+		$mofile_preferred = apply_filters( 'load_textdomain_mofile', $mofile_preferred, $domain );
+
+		/**
+		 * Filters the file path for loading translations for the given text domain.
+		 *
+		 * The file could be an MO or PHP file.
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param string $file   Path to the translation file to load.
+		 * @param string $domain The text domain.
+		 */
+		$mofile_preferred = apply_filters( 'load_translation_file', $mofile_preferred, $domain );
+
+		$success = WP_I18n_Translation_Controller::instance()->load( $mofile_preferred, $domain, $locale );
+
+		if ( $success ) {
+			if ( isset( $l10n[ $domain ] ) && $l10n[ $domain ] instanceof MO ) {
+				WP_I18n_Translation_Controller::instance()->load( $l10n[ $domain ]->get_filename(), $domain, $locale );
+			}
+
+			// Unset Noop_Translations reference in get_translations_for_domain.
+			unset( $l10n[ $domain ] );
+
+			$l10n[ $domain ] = new WP_I18n_Translations( $domain );
```

Comment:
So as far as I can tell, this `WP_I18n_Translations` class replaces the `MO` class? If that's the case, two related questions:
1. Is its interface fully compatible (to avoid potential BC breakage)?
2. Is there any remaining value for the `MO` class or could it effectively be deprecated?

Reply:
Yes it does.

1. Not 100%, but it covers the most common API touchpoints.
	- I haven't found any instances where more changes were required. Be it through tests or wpdirectory.net.
	- Actually extending `Translations` or the `MO` class is not possible because then the magic getter for `entries` and `headers` wouldn't work—unless we make changes to those classes.
	- I'd imagine getting to full compatibility gains only little safety for a lot of work.
2. Yes, there is. No, we will likely never be able to deprecate it. Many projects will continue to have MO files but no PHP files, and we need to support that. Plus, users getting deprecation warnings for no reason is not great. Second reason is that plugins like GlotPress directly use these classes.

Reply:
See also https://github.com/swissspidy/performant-translations/issues/43

Reply:
@swissspidy Not sure I agree on the 2nd point.
* Projects with only MO files can still handle them with the new class responsible for MO files.
* Deprecation warnings aren't for no reason. If there's code which is no longer used, it's a maintenance burden to keep it around. We don't delete code because that could introduce BC breakage, but deprecating unused code IMO is the right path. Plugins that use the `MO` class could gradually update to use `WP_I18n_Translations` / `WP_I18n_Translation_File_MO`.

Reply:
The MO class from the polo library as used by GlotPress can support many more things that are not covered by our class. I really don‘t think we can deprecate it. But let‘s ask @ocean90

Reply:
From what I can see, GlotPress mostly directly uses the `Translations` and `Translation_Entry` classes. These two and `MO` are all from the same pomo library, which is theoretically an external library. It would be weird to deprecate only one of those. So as an i18n component maintainer I am -1 on this suggestion.